explain analyze panic when insert select

## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
mysql> create table test (i int);
Query OK, 0 rows affected (0.01 sec)

mysql> explain analyze insert into test select 1;
ERROR 2013 (HY000): Lost connection to MySQL server during query
```

2. What did you expect to see?
I think insert is not supported by explain analyze, so it should return an error. Or we could explain the select part for this SQL.


3. What did you see instead?
Got a panic
```
[2019/07/02 20:32:34.950 +08:00] [ERROR] [conn.go:599] ["connection running loop panic"] [conn=390] [lastCmd="explain analyze insert into t1 select 1"] [err="\"invalid memory address or nil pointer dereference\""] [stack="goroutine 13837 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x56d4040, 0xc060035020, 0xc000193ad0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:597 +0xee\npanic(0x52076e0, 0x6742740)\n\t/usr/local/Cellar/go/1.12.6/libexec/src/runtime/panic.go:522 +0x1b5\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset.func1(0x0, 0x56e6220, 0xc05afd6460, 0xc003fa3778, 0x56d4040, 0xc060035020, 0xc000193ad0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:1255 +0x5a9\npanic(0x52076e0, 0x6742740)\n\t/usr/local/Cellar/go/1.12.6/libexec/src/runtime/panic.go:522 +0x1b5\ngithub.com/pingcap/tidb/session.(*TxnState).SetOption(0xc1425c32d0, 0x1, 0x0, 0x0)\n\t<autogenerated>:1 +0x2e\ngithub.com/pingcap/tidb/executor.(*InsertValues).addRecord(0xc0001528c0, 0xc0a5cdc840, 0x1, 0x1, 0xc1425c32d0, 0x0, 0xc03d1b4ba0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/insert_common.go:582 +0x221\ngithub.com/pingcap/tidb/executor.(*InsertExec).exec(0xc03d1b4b70, 0x56d4040, 0xc060035020, 0xc000696000, 0x1, 0x20, 0x0, 0x0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/insert.go:71 +0x167\ngithub.com/pingcap/tidb/executor.(*InsertValues).insertRowsFromSelect(0xc0001528c0, 0x56d4040, 0xc060035020, 0xc003fa33b8, 0x0, 0x0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/insert_common.go:339 +0x622\ngithub.com/pingcap/tidb/executor.(*InsertExec).Next(0xc03d1b4b70, 0x56d4040, 0xc060035020, 0xc03d1b4c60, 0x0, 0xc003fa34b0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/insert.go:139 +0xf4\ngithub.com/pingcap/tidb/executor.(*ExplainExec).generateExplainInfo(0xc05b5d00b0, 0x56d4040, 0xc060035020, 0xc003fa3470, 0x60, 0x70, 0x40e2748, 0xc0000c6008)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/explain.go:77 +0x154\ngithub.com/pingcap/tidb/executor.(*ExplainExec).Next(0xc05b5d00b0, 0x56d4040, 0xc060035020, 0xc03d1b4c00, 0x4, 0xc09a8fff18)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/explain.go:52 +0x42e\ngithub.com/pingcap/tidb/executor.Next(0x56d4040, 0xc060035020, 0x56da380, 0xc05b5d00b0, 0xc03d1b4c00, 0x0, 0x0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/executor.go:198 +0x119\ngithub.com/pingcap/tidb/executor.(*recordSet).Next(0xc05afd6410, 0x56d4040, 0xc060035020, 0xc03d1b4c00, 0xc05afd6410, 0xc03d1b4c00)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/adapter.go:104 +0x5a\ngithub.com/pingcap/tidb/server.(*tidbResultSet).Next(0xc05afd6460, 0x56d4040, 0xc060035020, 0xc03d1b4c00, 0x4, 0x400)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/driver_tidb.go:365 +0x51\ngithub.com/pingcap/tidb/server.(*clientConn).writeChunks(0xc000193ad0, 0x56d4040, 0xc060035020, 0x56e6220, 0xc05afd6460, 0xc000003700, 0x56d4040, 0xc060035020)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:1302 +0x319\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset(0xc000193ad0, 0x56d4040, 0xc060035020, 0x56e6220, 0xc05afd6460, 0xc000001f00, 0x0, 0x0, 0x0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:1268 +0x1a1\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc000193ad0, 0x56d4040, 0xc060035020, 0xc04239ec61, 0x27, 0x0, 0x0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:1185 +0x212\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch(0xc000193ad0, 0x56d4040, 0xc060035020, 0xc04239ec61, 0x28, 0x28, 0x0, 0x0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:895 +0x5c6\ngithub.com/pingcap/tidb/server.(*clientConn).Run(0xc000193ad0, 0x56d4040, 0xc060035020)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:650 +0x258\ngithub.com/pingcap/tidb/server.(*Server).onConn(0xc0007c7280, 0xc000193ad0)\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/server.go:441 +0x481\ncreated by github.com/pingcap/tidb/server.(*Server).Run\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/server.go:358 +0x75d\n"] [stack="github.com/pingcap/tidb/server.(*clientConn).Run.func1\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:599\nruntime.gopanic\n\t/usr/local/Cellar/go/1.12.6/libexec/src/runtime/panic.go:522\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset.func1\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:1255\nruntime.gopanic\n\t/usr/local/Cellar/go/1.12.6/libexec/src/runtime/panic.go:522\nruntime.panicmem\n\t/usr/local/Cellar/go/1.12.6/libexec/src/runtime/panic.go:82\nruntime.sigpanic\n\t/usr/local/Cellar/go/1.12.6/libexec/src/runtime/signal_unix.go:390\ngithub.com/pingcap/tidb/session.(*TxnState).SetOption\n\t<autogenerated>:1\ngithub.com/pingcap/tidb/executor.(*InsertValues).addRecord\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/insert_common.go:582\ngithub.com/pingcap/tidb/executor.(*InsertExec).exec\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/insert.go:71\ngithub.com/pingcap/tidb/executor.(*InsertValues).insertRowsFromSelect\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/insert_common.go:339\ngithub.com/pingcap/tidb/executor.(*InsertExec).Next\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/insert.go:139\ngithub.com/pingcap/tidb/executor.(*ExplainExec).generateExplainInfo\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/explain.go:77\ngithub.com/pingcap/tidb/executor.(*ExplainExec).Next\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/explain.go:52\ngithub.com/pingcap/tidb/executor.Next\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/executor.go:198\ngithub.com/pingcap/tidb/executor.(*recordSet).Next\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/executor/adapter.go:104\ngithub.com/pingcap/tidb/server.(*tidbResultSet).Next\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/driver_tidb.go:365\ngithub.com/pingcap/tidb/server.(*clientConn).writeChunks\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:1302\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:1268\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:1185\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:895\ngithub.com/pingcap/tidb/server.(*clientConn).Run\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/conn.go:650\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/Users/yusp/work/goprojects/src/github.com/pingcap/tidb/server/server.go:441"]
```


4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
ff82b62b973915b0820d1bfcec2e4aa0a1986437

--- a/executor/adapter.go
+++ b/executor/adapter.go
@@ -253,17 +253,8 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 		return a.handlePessimisticSelectForUpdate(ctx, e)
 	}
 
-	// If the executor doesn't return any result to the client, we execute it without delay.
-	if e.Schema().Len() == 0 {
-		if isPessimistic {
-			return nil, a.handlePessimisticDML(ctx, e)
-		}
-		return a.handleNoDelayExecutor(ctx, e)
-	} else if proj, ok := e.(*ProjectionExec); ok && proj.calculateNoDelay {
-		// Currently this is only for the "DO" statement. Take "DO 1, @a=2;" as an example:
-		// the Projection has two expressions and two columns in the schema, but we should
-		// not return the result of the two expressions.
-		return a.handleNoDelayExecutor(ctx, e)
+	if result, err := a.handleNoDelay(ctx, e, isPessimistic); result != nil || err != nil {
+		return result, err
 	}
 
 	var txnStartTS uint64
@@ -279,6 +270,30 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 		stmt:       a,
 		txnStartTS: txnStartTS,
 	}, nil
+}
+
+func (a *ExecStmt) handleNoDelay(ctx context.Context, e Executor, isPessimistic bool) (sqlexec.RecordSet, error) {
+	toCheck := e
+	if explain, ok := e.(*ExplainExec); ok {
+		if explain.analyzeExec != nil {
+			toCheck = explain.analyzeExec
+		}
+	}
+
+	// If the executor doesn't return any result to the client, we execute it without delay.
+	if toCheck.Schema().Len() == 0 {
+		if isPessimistic {
+			return nil, a.handlePessimisticDML(ctx, e)
+		}
+		return a.handleNoDelayExecutor(ctx, e)
+	} else if proj, ok := toCheck.(*ProjectionExec); ok && proj.calculateNoDelay {
+		// Currently this is only for the "DO" statement. Take "DO 1, @a=2;" as an example:
+		// the Projection has two expressions and two columns in the schema, but we should
+		// not return the result of the two expressions.
+		return a.handleNoDelayExecutor(ctx, e)
+	}
+
+	return nil, nil
 }
 
 // getMaxExecutionTime get the max execution timeout value.

--- a/executor/adapter.go
+++ b/executor/adapter.go
@@ -253,7 +253,7 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 		return a.handlePessimisticSelectForUpdate(ctx, e)
 	}
 
-	if result, err := a.handleNoDelay(ctx, e, isPessimistic); result != nil || err != nil {
+	if handled, result, err := a.handleNoDelay(ctx, e, isPessimistic); handled {
 		return result, err
 	}
 
@@ -272,7 +272,7 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 	}, nil
 }
 
-func (a *ExecStmt) handleNoDelay(ctx context.Context, e Executor, isPessimistic bool) (sqlexec.RecordSet, error) {
+func (a *ExecStmt) handleNoDelay(ctx context.Context, e Executor, isPessimistic bool) (bool, sqlexec.RecordSet, error) {
 	toCheck := e
 	if explain, ok := e.(*ExplainExec); ok {
 		if explain.analyzeExec != nil {
@@ -283,17 +283,19 @@ func (a *ExecStmt) handleNoDelay(ctx context.Context, e Executor, isPessimistic 
 	// If the executor doesn't return any result to the client, we execute it without delay.
 	if toCheck.Schema().Len() == 0 {
 		if isPessimistic {
-			return nil, a.handlePessimisticDML(ctx, e)
+			return true, nil, a.handlePessimisticDML(ctx, e)
 		}
-		return a.handleNoDelayExecutor(ctx, e)
+		r, err := a.handleNoDelayExecutor(ctx, e)
+		return true, r, err
 	} else if proj, ok := toCheck.(*ProjectionExec); ok && proj.calculateNoDelay {
 		// Currently this is only for the "DO" statement. Take "DO 1, @a=2;" as an example:
 		// the Projection has two expressions and two columns in the schema, but we should
 		// not return the result of the two expressions.
-		return a.handleNoDelayExecutor(ctx, e)
+		r, err := a.handleNoDelayExecutor(ctx, e)
+		return true, r, err
 	}
 
-	return nil, nil
+	return false, nil, nil
 }
 
 // getMaxExecutionTime get the max execution timeout value.

--- a/executor/explain_test.go
+++ b/executor/explain_test.go
@@ -63,11 +63,14 @@ func (s *testSuite1) TestExplainPriviliges(c *C) {
 
 func (s *testSuite1) TestExplainWrite(c *C) {
 	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("drop table if exists t")
 	tk.MustExec("create table t (a int)")
-	tk.MustExec("explain analyze insert into t values (1)")
+	tk.MustExec("explain analyze insert into t select 1")
 	tk.MustQuery("select * from t").Check(testkit.Rows("1"))
 	tk.MustExec("explain analyze update t set a=2 where a=1")
 	tk.MustQuery("select * from t").Check(testkit.Rows("2"))
-	tk.MustExec("explain analyze insert into t values (1)")
+	tk.MustExec("explain insert into t select 1")
 	tk.MustQuery("select * from t").Check(testkit.Rows("2"))
+	tk.MustExec("explain analyze insert into t select 1")
+	tk.MustQuery("select * from t order by a").Check(testkit.Rows("1", "2"))
 }

--- a/executor/explain_test.go
+++ b/executor/explain_test.go
@@ -60,3 +60,14 @@ func (s *testSuite1) TestExplainPriviliges(c *C) {
 	err = tk1.ExecToErr("explain select * from v")
 	c.Assert(err.Error(), Equals, plannercore.ErrTableaccessDenied.GenWithStackByArgs("SELECT", "explain", "%", "v").Error())
 }
+
+func (s *testSuite1) TestExplainWrite(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("create table t (a int)")
+	tk.MustExec("explain analyze insert into t values (1)")
+	tk.MustQuery("select * from t").Check(testkit.Rows("1"))
+	tk.MustExec("explain analyze update t set a=2 where a=1")
+	tk.MustQuery("select * from t").Check(testkit.Rows("2"))
+	tk.MustExec("explain analyze insert into t values (1)")
+	tk.MustQuery("select * from t").Check(testkit.Rows("2"))
+}

--- a/go.mod
+++ b/go.mod
@@ -41,7 +41,7 @@ require (
 	github.com/pingcap/goleveldb v0.0.0-20171020122428-b9ff6c35079e
 	github.com/pingcap/kvproto v0.0.0-20190703131923-d9830856b531
 	github.com/pingcap/log v0.0.0-20190307075452-bd41d9273596
-	github.com/pingcap/parser v0.0.0-20190708123555-29973f7a22eb
+	github.com/pingcap/parser v0.0.0-20190710072914-6cd203114f2d
 	github.com/pingcap/pd v0.0.0-20190617100349-293d4b5189bf
 	github.com/pingcap/tidb-tools v2.1.3-0.20190321065848-1e8b48f5c168+incompatible
 	github.com/pingcap/tipb v0.0.0-20190428032612-535e1abaa330

